File: robotics-codes-from-scratch-master/matlab/LQR_infHor.m
%% Point-mass LQR with infinite horizon
%% 
%% Copyright (c) 2023 Idiap Research Institute <https://www.idiap.ch/>
%% Written by Sylvain Calinon <https://calinon.ch>
%% 
%% This file is part of RCFS <https://robotics-codes-from-scratch.github.io/>
%% License: MIT


function LQR_infHor
    
%% Parameters
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
param.nbData = 200; %Number of datapoints
param.nbRepros = 4; %Number of reproductions

param.nbVarPos = 2; %Dimension of position data (here: x1,x2)
param.nbDeriv = 2; %Number of static & dynamic features (D=2 for [x,dx])
param.nbVar = param.nbVarPos * param.nbDeriv; %Dimension of state vector in the tangent space
param.dt = 1E-2; %Time step duration
param.rfactor = 4E-2;	%Control cost in LQR 

%Control cost matrix
R = eye(param.nbVarPos) * param.rfactor;

%Target and desired covariance
param.Mu = [randn(param.nbVarPos,1); zeros(param.nbVarPos*(param.nbDeriv-1),1)];

[Ar,~] = qr(randn(param.nbVarPos));
xCov = Ar * diag(rand(param.nbVarPos,1)) * Ar' * 1E-1

%% Discrete dynamical System settings 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
A1d = zeros(param.nbDeriv);
for i=0:param.nbDeriv-1
    A1d = A1d + diag(ones(param.nbDeriv-i,1),i) * param.dt^i * 1/factorial(i); %Discrete 1D
end
B1d = zeros(param.nbDeriv,1); 
for i=1:param.nbDeriv
    B1d(param.nbDeriv-i+1) = param.dt^i * 1/factorial(i); %Discrete 1D
end
A = kron(A1d, eye(param.nbVarPos)); %Discrete nD
B = kron(B1d, eye(param.nbVarPos)); %Discrete nD


%% discrete LQR with infinite horizon
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
Q = blkdiag(inv(xCov), zeros(param.nbVarPos*(param.nbDeriv-1))); %Precision matrix
P = solveAlgebraicRiccati_eig_discrete(A, B*(R\B'), (Q+Q')/2);
L = (B' * P * B + R) \ B' * P * A; %Feedback gain (discrete version)

for n=1:param.nbRepros
    x = [ones(param.nbVarPos,1)+randn(param.nbVarPos,1)*5E-1; zeros(param.nbVarPos*(param.nbDeriv-1),1)];
    for t=1:param.nbData		
        r(n).Data(:,t) = x; 
        u = L * (param.Mu - x); %Compute acceleration (with only feedback terms)
        x = A * x + B * u;
    end
end


%% Plots
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


figure('position',[10,10,650,650]); hold on; axis off; grid off; 
plotGMM(param.Mu(1:2), xCov(1:2,1:2), [.8 0 0], .3);
for n=1:param.nbRepros
    plot(r(n).Data(1,:), r(n).Data(2,:), '-','linewidth',1,'color',[0 0 0]);
end
%plot(param.Mu(1,1), param.Mu(2,1), 'r.','markersize',80);
axis equal; 
%print('-dpng','graphs/demo_MPC_infHor01.png');


%% Timeline plot
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

labList = {'x_1','x_2','dx_1','dx_2','ddx_1','ddx_2'};
figure('position',[720 10 600 650],'color',[1 1 1]); 
for j=1:param.nbVar
subplot(param.nbVar+1,1,j); hold on;
for n=1:param.nbRepros
    plot(r(n).Data(j,:), '-','linewidth',1,'color',[0 0 0]);
end
if j<7
    ylabel(labList{j},'fontsize',14,'interpreter','latex');
end
end
%Speed profile
if param.nbDeriv>1
subplot(param.nbVar+1,1,param.nbVar+1); hold on;
for n=1:param.nbRepros
    sp = sqrt(r(n).Data(3,:).^2 + r(n).Data(4,:).^2);
    plot(sp, '-','linewidth',1,'color',[0 0 0]);
end
ylabel('|dx|','fontsize',14);
xlabel('t','fontsize',14);
end
%print('-dpng','graphs/demo_LQR_infHor01.png');

pause;
close all;
end

%%%%%%%%%%%%%%%%%%%%%%
% Solve Algebraic Ricatty discrete equation
function X = solveAlgebraicRiccati_eig_discrete(A, G, Q)
    n = size(A,1);

    %Symplectic matrix (see https://en.wikipedia.org/wiki/Algebraic_Riccati_equation)
    %Z = [A+B*(R\B')/A'*Q, -B*(R\B')/A'; -A'\Q, A'^-1]; 
    Z = [A+G/A'*Q, -G/A'; -A'\Q, inv(A')];


    %Since Z is symplectic, if it does not have any eigenvalues on the unit circle, 
    %then exactly half of its eigenvalues are inside the unit circle. 
    [V,D] = eig(Z);
    U = [];
    for j=1:2*n
        if norm(D(j,j)) < 1 %inside unit circle
            U = [U V(:,j)];
        end
    end

    X = real(U(n+1:end,:) / U(1:n,:));
end

%%%%%%%%%%%%%%%%%%%%%%
% Plot a 2D Mixture of Gaussians
function [h, X] = plotGMM(Mu, Sigma, color, valAlpha)
    nbStates = size(Mu,2);
    nbDrawingSeg = 100;
    darkcolor = color * .7; %max(color-0.5,0);
    t = linspace(-pi, pi, nbDrawingSeg);
    if nargin<4
        valAlpha = 1;
    end

    h = [];
    X = zeros(2,nbDrawingSeg,nbStates);
    for i=1:nbStates
        [V,D] = eig(Sigma(:,:,i));
        R = real(V*D.^.5);
        X(:,:,i) = R * [cos(t); sin(t)] + repmat(Mu(:,i), 1, nbDrawingSeg);
        if nargin>3 %Plot with alpha transparency
            h = [h patch(X(1,:,i), X(2,:,i), color, 'lineWidth', 1, 'EdgeColor', color, 'facealpha', valAlpha,'edgealpha', valAlpha)];
            h = [h plot(Mu(1,:), Mu(2,:), '.', 'markersize', 10, 'color', darkcolor)];
        else %Plot without transparency
            h = [h patch(X(1,:,i), X(2,:,i), color, 'lineWidth', 1, 'EdgeColor', darkcolor)];
            h = [h plot(Mu(1,:), Mu(2,:), '.', 'markersize', 10, 'color', darkcolor)];
        end
    end
end

File: robotics-codes-from-scratch-master/python/LQR_infHor.py
'''
Point-mass LQR with infinite horizon

Copyright (c) 2023 Idiap Research Institute <https://www.idiap.ch/>
Written by Jérémy Maceiras <jeremy.maceiras@idiap.ch> and
Sylvain Calinon <https://calinon.ch>

This file is part of RCFS <https://robotics-codes-from-scratch.github.io/>
License: MIT
'''

import numpy as np
import copy
from scipy.linalg import solve_discrete_are as solve_algebraic_riccati_discrete
from scipy.stats import multivariate_normal
from math import factorial
import matplotlib.pyplot as plt

# Plot a 2D Gaussians
def plot_gaussian(mu, sigma):
    w, h = 100, 100

    std = [np.sqrt(sigma[0, 0]), np.sqrt(sigma[1, 1])]
    x = np.linspace(mu[0] - 3 * std[0], mu[0] + 3 * std[0], w)
    y = np.linspace(mu[1] - 3 * std[1], mu[1] + 3 * std[1], h)

    x, y = np.meshgrid(x, y)

    x_ = x.flatten()
    y_ = y.flatten()
    xy = np.vstack((x_, y_)).T

    normal_rv = multivariate_normal(mu, sigma)
    z = normal_rv.pdf(xy)
    z = z.reshape(w, h, order='F')

    plt.contourf(x, y, z.T,levels=1,colors=["white","red"],alpha=.5)

# Parameters
# ===============================

param = lambda: None # Lazy way to define an empty class in python
param.nbData = 200 # Number of datapoints
param.nbRepros = 4 #Number of reproductions
param.nbVarPos = 2 #Dimension of position data (here: x1,x2)
param.nbDeriv = 2 #Number of static & dynamic features (D=2 for [x,dx])
param.nbVar = param.nbVarPos * param.nbDeriv #Dimension of state vector in the tangent space
param.dt = 1e-2 #Time step duration
param.rfactor = 4e-2	#Control cost in LQR 

# Control cost matrix
R = np.identity(param.nbVarPos) * param.rfactor

# Target and desired covariance
param.Mu = np.hstack(( np.random.uniform(size=param.nbVarPos) , np.zeros(param.nbVarPos) ))
Ar,_ = np.linalg.qr(np.random.uniform(size=(param.nbVarPos,param.nbVarPos)))
xCov = Ar @ np.diag(np.random.uniform(size=param.nbVarPos)) @ Ar.T * 1e-1

# Discrete dynamical System settings 
# ===================================

A1d = np.zeros((param.nbDeriv,param.nbDeriv))
B1d = np.zeros((param.nbDeriv,1))

for i in range(param.nbDeriv):
    A1d += np.diag( np.ones(param.nbDeriv-i) ,i ) * param.dt**i * 1/factorial(i)
    B1d[param.nbDeriv-i-1] = param.dt**(i+1) * 1/factorial(i+1)

A = np.kron(A1d,np.identity(param.nbVarPos))
B = np.kron(B1d,np.identity(param.nbVarPos))

# Discrete LQR with infinite horizon
# ===================================

Q = np.zeros((param.nbVar,param.nbVar))
Q[:param.nbVarPos,:param.nbVarPos] = np.linalg.inv(xCov) # Precision matrix
P = solve_algebraic_riccati_discrete(A,B,Q,R)
L = np.linalg.inv(B.T @ P @ B + R) @ B.T @ P @ A # Feedback gain (discrete version)

reproducitons = []

for i in range(param.nbRepros):
    xt = np.zeros(param.nbVar)
    xt[:param.nbVarPos] = 1+np.random.uniform(param.nbVarPos)*2 
    xs = [copy.deepcopy(xt)]
    for t in range(param.nbData):
        u = L @ (param.Mu - xt)
        xt = A @ xt + B @ u
        xs += [copy.deepcopy(xt)]
    reproducitons += [ np.asarray(xs) ]

# Plots
# ======

plt.figure()
plt.title("Position")

for r in reproducitons:
    plt.plot(r[:,0],r[:,1],c="black")
plot_gaussian(param.Mu[:param.nbVarPos],xCov)
plt.axis("off")
plt.gca().set_aspect('equal', adjustable='box')

fig,axs = plt.subplots(5)

for r in reproducitons:
    axs[0].plot(r[:,0],c="black",alpha=.4,linestyle="dashed")
    axs[1].plot(r[:,1],c="black",alpha=.4,linestyle="dashed")    
    axs[2].plot(r[:,2],c="black",alpha=.4,linestyle="dashed")
    axs[3].plot(r[:,3],c="black",alpha=.4,linestyle="dashed")
    axs[4].plot(np.linalg.norm(r[:,2:4],axis=1),c="black",alpha=.4,linestyle="dashed")

axs[0].set_ylabel("$x_1$")
axs[1].set_ylabel("$x_2$")
axs[2].set_ylabel("$\dot{x}_1$")
axs[3].set_ylabel("$\dot{x}_2$")
axs[4].set_ylabel("$| \dot{x} |$")
plt.xlabel("T")

plt.show()
